core/src/tdx.rs

```diff
@@ -0,0 +1,410 @@
+            last_payload = event.payload_as_string();
+        }
+    }
+
+    if let Some(payload) = last_payload {
+        if payload == cert_hash {
+            return Ok(());
+        }
+    }
+
+    Err(RatlsError::Policy(
+        "TLS public key hash missing from event log".into(),
+    ))
+}
+
+impl TcgEvent {
+    fn digest_bytes(&self) -> Result<Vec<u8>, RatlsError> {
+        decode_hex_field(&self.digest)
+            .map_err(|e| RatlsError::Vendor(format!("invalid digest for event '{}': {e}", self.label())))
+    }
+
+    fn payload_as_string(&self) -> Option<String> {
+        let raw = self.event_payload.trim();
+        if raw.is_empty() {
+            return None;
+        }
+        let bytes = decode_hex_field(raw).ok()?;
+        String::from_utf8(bytes).ok()
+    }
+
+    fn label(&self) -> String {
+        let trimmed = self.event.trim();
+        if trimmed.is_empty() {
+            format!("type {}", self.event_type)
+        } else {
+            trimmed.to_string()
+        }
+    }
+}
+
+fn enforce_tcb_policy(report: &TDReport10, policy: &TdxTcbPolicy) -> Result<(), RatlsError> {
+    if let Some(expected) = &policy.mrseam {
+        if report.mr_seam[..expected.len()] != expected[..] {
+            return Err(RatlsError::Policy("mr_seam mismatch".into()));
+        }
+    }
+    if let Some(expected) = &policy.mrtmrs {
+        if report.rt_mr0[..expected.len()] != expected[..] {
+            return Err(RatlsError::Policy("rt_mr0 mismatch".into()));
+        }
+    }
+    Ok(())
+}
+
+enum TdReportRef<'a> {
+    Td10(&'a TDReport10),
+    Td15(&'a TDReport15),
+}
+
+impl<'a> TdReportRef<'a> {
+    fn from_report(report: &'a Report) -> Result<Self, RatlsError> {
+        match report {
+            Report::TD10(r) => Ok(TdReportRef::Td10(r)),
+            Report::TD15(r) => Ok(TdReportRef::Td15(r)),
+            other => Err(RatlsError::TeeUnsupported(format!(
+                "unsupported report type: {other:?}"
+            ))),
+        }
+    }
+
+    fn as_td10(&self) -> &TDReport10 {
+        match self {
+            TdReportRef::Td10(r) => r,
+            TdReportRef::Td15(r) => &r.base,
+        }
+    }
+
+    fn report_data(&self) -> &[u8; 64] {
+        match self {
+            TdReportRef::Td10(r) => &r.report_data,
+            TdReportRef::Td15(r) => &r.base.report_data,
+        }
+    }
+}
+
+fn parse_quote_report(quote: &[u8]) -> Result<Quote, RatlsError> {
+    Quote::parse(quote).map_err(|e| RatlsError::Vendor(format!("failed to parse quote: {e}")))
+}
+
+fn decode_hex_field(value: &str) -> Result<Vec<u8>, String> {
+    let normalized = normalize_hex(value);
+    decode(&normalized).map_err(|e| format!("invalid hex '{value}': {e}"))
+}
+
+fn normalize_hex(value: &str) -> String {
+    let mut lowered = value.trim().to_ascii_lowercase();
+    if lowered.starts_with("0x") {
+        lowered.drain(..2);
+    }
+    lowered.retain(|ch| !ch.is_ascii_whitespace());
```

Comment:
IMHO, we should either clean it by selecting hex chars (so ignoring whitespaces or any other invalid char), or do nothing. I don't see the point of removing whitespaces only as a whitespace and a 'Z' for example are the same in terms of hex encoding: they are both invalid chars.